python/dgl/distributed/client.py

```diff
@@ -2,33 +2,155 @@
 
 """Functions used by client."""
 
-def connect_to_server(server_namebook):
+from . import rpc
+
+import os
+import time
+import socket
+
+if os.name != 'nt':
+    import fcntl
+    import struct
+
+def local_ip4_addr_list():
+    """Return a set of IPv4 address
+    """
+    nic = set()
+    for if_nidx in socket.if_nameindex():
+        name = if_nidx[1]
+        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        ip_addr = socket.inet_ntoa(fcntl.ioctl(
+            sock.fileno(),
+            0x8915,  # SIOCGIFADDR
+            struct.pack('256s', name[:15].encode("UTF-8")))[20:24])
+        nic.add(ip_addr)
+    return nic
+
+def get_local_machine_id(server_namebook):
+    """Given server_namebook, find local machine ID
+
+    Parameters
+    ----------
+    server_namebook: dict
+        IP address namebook of server nodes, where key is the server's ID
+        (start from 0) and value is the server's machine_id, IP address,
+        port, and group_count, e.g.,
+
+          {0:'[0, 172.31.40.143, 30050, 2],
+           1:'[0, 172.31.40.143, 30051, 2],
+           2:'[1, 172.31.36.140, 30050, 2],
+           3:'[1, 172.31.36.140, 30051, 2],
+           4:'[2, 172.31.47.147, 30050, 2],
+           5:'[2, 172.31.47.147, 30051, 2],
+           6:'[3, 172.31.30.180, 30050, 2],
+           7:'[3, 172.31.30.180, 30051, 2]}
+
+    Returns
+    -------
+    int
+        local machine ID
+    """
+    res = 0
+    for ID, data in server_namebook.items():
+        machine_id = data[0]
+        ip = data[1]
+        if ip in local_ip4_addr_list():
+            res = machine_id
+            break
+    return res
+
+def get_local_usable_addr():
+    """Get local usable IP and port
+
+    Returns
+    -------
+    str
+        IP address, e.g., '192.168.8.12:50051'
+    """
+    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    try:
+        # doesn't even have to be reachable
+        s.connect(('10.255.255.255', 1))
+        ip_addr = s.getsockname()[0]
+    except:
+        ip_addr = '127.0.0.1'
+    finally:
+        s.close()
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s.bind(("", 0))
+    s.listen(1)
+    port = s.getsockname()[1]
+    s.close()
+
+    return ip_addr + ':' + str(port)
+
+def connect_to_server(ip_config, queue_size=20*1024*1024*1024, net_type='socket'):
     """Connect this client to server.
 
     Parameters
     ----------
-    server_namebook : ...
-        Server names.
+    ip_config : str
+        Path of server IP configuration file.
+    queue_size : int
+        Size (bytes) of client queue buffer (~20 GB on default).
+        Note that the 20 GB is just an upper-bound and DGL uses zero-copy and
+        it will not allocate 20GB memory at once.
+    net_type : str
+        networking type, e.g., 'socket' (on default) or 'mpi' (do not support yet).
 
     Raises
     ------
     ConnectionError : If anything wrong with the connection.
```

Comment:
Have you properly handled exceptions?

Reply:
The exception will be catched by C++ code.